### DISCUSSION TITLE
[Chapter 03] IllegalArgumentException vs. IllegalStateException

### DISCUSSION THREAD

**Discussion by corock** [cowaine/spring-boot-for-msa#13] [Chapter 03. Spring Application Basics] [answered]
_p. 108_ `PriceUnit` 클래스를 보면 생성자와 `validate()` 메서드에 각각 `null` 체크를 하고 있습니다.
그런데 예외 타입이 [IllegalArgumentException](https://github.com/cowaine/spring-boot-for-msa/blob/feature/corock/corock/src/main/java/com/cowaine/corock/chapter03/domain/PriceUnit.java#L19) 과 [IllegalStateException](https://github.com/cowaine/spring-boot-for-msa/blob/feature/corock/corock/src/main/java/com/cowaine/corock/chapter03/domain/PriceUnit.java#L32) 클래스인데, 왜 각각 다르게 던지는 건가요?

**Comment by JoisFe** [comment, marked as answer]
@corock 님께서 예전에 정리해주신 글 참고하여 다시 한번 상기하게 되었네요

### `IllegalArgumentException`

> 호출자가 인수로 부적절한 값을 넘길 때 던지는 예외

-  e.g. 반복 횟수를 지정하는 매개변수에 음수를 전달할 때

### `IllegalStateException`

> 대상 객체의 상태가 호출된 메서드를 수행하기에 적합하지 않을 때 던지는 예외

- e.g. 제대로 초기화되지 않은 객체를 사용하려 할 때

### [참고자료](https://github.com/orgs/Study-2-Effective-Java/discussions/175)

**Comment by Irisation23** [comment]
<img width="873" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/53285909/972f2764-9a42-4ada-b8fb-09bebd63e462">

- https://yangbongsoo.gitbook.io/study/oo_fact_and_misunderstanding
